File: windows/Services/TransferHistory.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace BluedropWindows.Services;

/// <summary>One completed transfer, shown in the history window.</summary>
public record TransferRecord
{
    [JsonPropertyName("id")] public string Id { get; init; } = Guid.NewGuid().ToString("N");
    [JsonPropertyName("timestamp")] public long Timestamp { get; init; } // epoch ms
    [JsonPropertyName("direction")] public string Direction { get; init; } = "sent"; // sent|received
    [JsonPropertyName("kind")] public string Kind { get; init; } = "file"; // file|image
    [JsonPropertyName("name")] public string Name { get; init; } = "";
    [JsonPropertyName("size")] public long Size { get; init; }
    [JsonPropertyName("path")] public string? Path { get; init; } // on-disk payload; null for sent/text items
    [JsonPropertyName("mime")] public string Mime { get; init; } = "application/octet-stream";
    [JsonPropertyName("status")] public string Status { get; init; } = "ok"; // ok|failed; default keeps old files readable
    [JsonPropertyName("error")] public string? Error { get; init; }
}

/// <summary>
/// Transfer history persisted to %APPDATA%\Bluedrop\history.json. Records
/// file/image/text transfers in both directions, including failed/cancelled
/// ones (Status = "failed").
/// </summary>
public static class TransferHistory
{
    private const int MaxRecords = 500;

    private static readonly string FilePath =
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Bluedrop", "history.json");

    private static readonly object Gate = new();
    private static List<TransferRecord> _records = [];

    /// <summary>Raised on the calling thread whenever the list changes.</summary>
    public static event Action? Changed;

    public static IReadOnlyList<TransferRecord> Records
    {
        get { lock (Gate) { return _records.ToList(); } }
    }

    public static void Load()
    {
        lock (Gate)
        {
            _records = LoadLocked();
        }
        Changed?.Invoke();
    }

    public static void Add(TransferRecord record)
    {
        lock (Gate)
        {
            _records = new[] { record }.Concat(_records).Take(MaxRecords).ToList();
            SaveLocked();
        }
        Changed?.Invoke();
    }

    /// <summary>Removes the record, keeping any stored payload.</summary>
    public static void Remove(string id)
    {
        lock (Gate)
        {
            _records = _records.Where(r => r.Id != id).ToList();
            SaveLocked();
        }
        Changed?.Invoke();
    }

    /// <summary>Removes the record and deletes its stored payload.</summary>
    public static bool RemoveWithFile(string id, out string? failedPath)
    {
        TransferRecord? record;
        lock (Gate) { record = _records.FirstOrDefault(r => r.Id == id); }
        failedPath = null;
        var deleted = true;
        if (record?.Path is { Length: > 0 } path)
        {
            try { if (File.Exists(path)) File.Delete(path); }
            catch { deleted = false; failedPath = path; }
        }
        Remove(id);
        return deleted;
    }

    public static void Clear()
    {
        lock (Gate)
        {
            _records = [];
            SaveLocked();
        }
        Changed?.Invoke();
    }

    private static List<TransferRecord> LoadLocked()
    {
        try
        {
            if (!File.Exists(FilePath)) return [];
            using var doc = JsonDocument.Parse(File.ReadAllBytes(FilePath));
            return doc.RootElement.EnumerateArray()
                .Select(e => e.Deserialize<TransferRecord>())
                .Where(r => r != null)
                .Cast<TransferRecord>()
                .ToList();
        }
        catch
        {
            return [];
        }
    }

    private static void SaveLocked()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(FilePath)!);
            File.WriteAllText(FilePath, JsonSerializer.Serialize(_records));
        }
        catch
        {
            // history is best-effort; never break transfers over it
        }
    }
}

File: android/app/src/main/java/com/bluedrop/core/TransferHistory.kt
package com.bluedrop.core

import android.content.Context
import android.net.Uri
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Transfer history shown in the app's History screen. Records file/image/text
 * transfers in both directions, including failed/cancelled ones ([status] =
 * "failed"). One JSON file in app-private storage, capped at [MAX_RECORDS].
 */
@Serializable
data class TransferRecord(
    val id: String,
    val timestamp: Long,
    val direction: String, // "sent" | "received"
    val kind: String,      // "file" | "image" | "text"
    val name: String,
    val size: Long,
    val uri: String? = null, // content:// of the stored payload; null for sent/text items
    val mime: String = "application/octet-stream",
    val status: String = "ok", // "ok" | "failed"; default keeps old JSON files readable
    val error: String? = null,
)

object TransferHistory {
    private const val MAX_RECORDS = 500
    private const val FILE_NAME = "transfer_history.json"

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val _records = MutableStateFlow<List<TransferRecord>>(emptyList())
    val records: StateFlow<List<TransferRecord>> = _records.asStateFlow()

    private var storeFile: File? = null
    private val lock = Any()

    fun init(context: Context) {
        synchronized(lock) {
            storeFile = File(context.filesDir, FILE_NAME)
            _records.value = runCatching {
                json.decodeFromString<List<TransferRecord>>(storeFile!!.readText())
            }.getOrDefault(emptyList())
        }
    }

    fun add(record: TransferRecord) {
        synchronized(lock) {
            _records.value = (listOf(record) + _records.value).take(MAX_RECORDS)
            persistLocked()
        }
    }

    /** Removes the record from the list, keeping any stored payload. */
    fun remove(id: String) {
        synchronized(lock) {
            _records.value = _records.value.filterNot { it.id == id }
            persistLocked()
        }
    }

    /**
     * Removes the record and deletes its stored payload (received files only).
     * Returns false when the payload could not be deleted; the record is
     * removed regardless.
     */
    fun removeWithFile(context: Context, id: String): Boolean {
        val record = synchronized(lock) { _records.value.firstOrNull { it.id == id } }
        var fileDeleted = false
        if (record?.uri != null) {
            fileDeleted = runCatching {
                context.contentResolver.delete(Uri.parse(record.uri), null, null) > 0
            }.getOrDefault(false)
        }
        remove(id)
        return fileDeleted
    }

    fun clear() {
        synchronized(lock) {
            _records.value = emptyList()
            persistLocked()
        }
    }

    private fun persistLocked() {
        val file = storeFile ?: return
        runCatching {
            val tmp = File(file.parentFile, "$FILE_NAME.tmp")
            tmp.writeText(json.encodeToString(_records.value))
            if (!tmp.renameTo(file)) {
                file.writeText(json.encodeToString(_records.value))
                tmp.delete()
            }
        }
    }
}
